src/microbe_stage/editor/EditorActionHistory.cs

```diff
@@ -0,0 +1,110 @@
+        foreach (var historyAction in cache ??= GetActionHistorySinceLastNewMicrobePress())
+        {
+            switch (action.GetInterferenceModeWith(historyAction))
+            {
+                case MicrobeActionInterferenceMode.Combinable:
+                    result -= actionCost;
+                    result -= historyAction.CalculateCost();
+                    result += action.Combine(historyAction).CalculateCost();
+                    break;
+                case MicrobeActionInterferenceMode.CancelsOut:
+                    result -= actionCost;
+                    result -= historyAction.CalculateCost();
+                    return result;
+                case MicrobeActionInterferenceMode.ReplacesOther:
+                    result -= historyAction.CalculateCost();
+                    break;
+            }
+        }
+
+        return result;
+    }
+
+    /// <summary>
+    ///   Calculates the remaining MP from the action history
+    /// </summary>
+    /// <returns>The remaining MP</returns>
+    public int CalculateMutationPointsLeft()
+    {
+        var copyLength = (cache ??= GetActionHistorySinceLastNewMicrobePress()).Count;
+        for (int compareToIndex = 0; compareToIndex < copyLength - 1; compareToIndex++)
+        {
+            for (int compareIndex = compareToIndex + 1; compareIndex < copyLength; compareIndex++)
+            {
+                switch (cache[compareIndex].GetInterferenceModeWith(cache[compareToIndex]))
+                {
+                    case MicrobeActionInterferenceMode.NoInterference:
+                        break;
+                    case MicrobeActionInterferenceMode.Combinable:
+                        var combinedValue = cache[compareIndex].Combine(cache[compareToIndex]);
+                        cache.RemoveAt(compareIndex);
+                        cache.RemoveAt(compareToIndex);
+                        compareIndex--;
+                        cache.Insert(compareIndex--, combinedValue);
+                        copyLength--;
+                        break;
+                    case MicrobeActionInterferenceMode.ReplacesOther:
+                        cache.RemoveAt(compareToIndex);
+                        compareIndex = compareToIndex;
+                        copyLength--;
+                        break;
+                    case MicrobeActionInterferenceMode.CancelsOut:
+                        cache.RemoveAt(compareIndex);
+                        cache.RemoveAt(compareToIndex);
+                        compareIndex -= 2;
+                        copyLength -= 2;
+                        break;
+                    default:
+                        throw new NotImplementedException();
+                }
+            }
+        }
+
+        return Constants.BASE_MUTATION_POINTS - cache.Sum(p => p.CalculateCost());
+    }
+
+    public override bool Redo()
+    {
+        var result = base.Redo();
+        cache = GetActionHistorySinceLastNewMicrobePress();
```

Comment:
Is this necessary? Can we just add an action here?

Reply:
It isn't really possible to just remove the last action because it might already have been reduced down in the `CalculateMutationPointsLeft` method. Adding a "Redo" action might be possible, but this further complicates the already way too complicated code